### PR TITLE
Inlude package-lock.json

### PR DESCRIPTION
**Pull request by TimDaub**
See: https://stackoverflow.com/a/44210813/1263876

Added package-lock.json. If you're preferring yarn, `yarn.lock` is also supposed to be committed.
Not sure how to keep both up to date at the same time 🤷‍♂️ 

Input welcome.

### CODE REVIEW

**Comment by dmihal** [comment]
Including `package-lock.json` makes sense.

It's not easy to keep `package-lock` and `yarn.lock` synced, the project should stick with either NPM or Yarn.

**Comment by dbe** [comment]
@TimDaub @dmihal @austintgriffith 

Lets keep an eye on this one. We used to commit package.lock but there are some really infuriating issues related to peer dependencies which create this circular loop of changing dependencies. It sounds convoluted, but I wasted a good 8+ hours hunting it down and it is a known issue related to contentious design decisions made by the npm team.

The way it happens is that you will clone the repo and npm install. This will actually change the package-lock file (not great). Then if you delete your node_modules folder and npm install again, it will change the package-lock *back*. It is insane and there are many long long threads on npm forums/github issues trying to isolate these issues.

TLDR: If you see crazy flip-flopping package-lock.json issues, I suggest we go back to just taking it out of source control until peer dependencies settle down or npm changes this behavior.

**Comment by TimDaub** [comment]
Hi Brian,

I proposed this change because we had three people at our team debug an issue for 3-4hrs yesterday as we couldn't figure out what went wrong with our development. The following happened:

We all developed on the burner-wallet but simply expected package-lock.json to be checked-in into the repository. As a new package-lock.json was created on each of our machines respectively and since package-lock.json takes precedent over package.json when installing dependencies, we ended up with different build outputs.
This lead to the burner wallet working completely different on all of our machines. For some devs it worked, for some devs everything broke. Only when we discovered that package-lock.json wasn't checked-in to the repository we knew what was going on.

To address your points, specifically

> The way it happens is that you will clone the repo and npm install. This will actually change the package-lock file (not great). Then if you delete your node_modules folder and npm install again, it will change the package-lock back.

I'd recommend to pin the dependencies in package.json and manually update them. Additionally, you can use a service like [greenkeeper](https://greenkeeper.io/) to keep them up to date. Not only will this stop package-lock.json from being changed, it is also [vital for security](https://www.theregister.co.uk/2018/11/26/npm_repo_bitcoin_stealer/).

I'm also not a big fan of package-lock.json but when working in teams with npm it simply cannot be ignored.

**Comment by austintgriffith** [comment]
Yeah I'm starting to see this issue pop up now where it wants to go back and forth between package lock versions: 
image

We might have to remove the package-lock if this continues

**Comment by TimDaub** [comment]
Not sure what you did, but package-lock.json is to be commited everytime it changes (at least 99% of cases).
It changes when e.g.:

- When you install a new dependency
- When a dependency updates upstream; As you have all of your package.json dependencies set to `^`, this might happen very often. That's why I was suggesting to pin them (remove `^` in front) and setup greenkeeper.

In your case it's likely you'll just have to do:

```
git add package-lock.json
git commit -m "Update package-lock.json"
git checkout <new branch>
```

I cannot recommend removing package-lock.json, as it often takes precedence over package.json and makes sure you get the package you expect. Especially, when working as a team on a code base.

**Comment by dbe** [comment]
@TimDaub I totally agree with your sentiment. I was in your position a month or so ago trying to get package-lock added to the repo. I strongly believe in the concept, especially on teams of any size really.

This is a forum post outlining the issue that we ran into in regards to the "weird" circular behavior: [Link](https://npm.community/t/package-lock-json-is-not-complete-on-first-run-for-some-modules/1817)

I still don't fully 100% grok what is going on, but my best understanding is that there is some issue with optional and peer dependencies where it is possible to create this unintended behavior. Every time you npm install, it changes the package-lock *even though no upstream things were changed*. And even more amusing is that it many times creates a cycle, so after enough times of 
```
npm install
rm -rf node_modules
```

You end up back to the original package-lock.json


This manifests itself with developers doing "fresh installs" which is why it is so frustrating. 

1) Person A clones the repo and runs "npm install"
2) package-lock.json changes (even though no upstream deps changed)
3) Person A commits the new package-lock.json and pushes to master
4) Person B clones the repo and runs "npm install"
5) package-lock.json changes again (once again no upstream deps changed)
.... goes on forever.


I still don't know what the proper solution is here, but we came to the conclusion before that getting rid of package-lock.json was the best bet. That may have been wrong, and things may now have changed.

Hope this shines a little light on some of the annoying issues of npm in regards to our project.

**Comment by mkosowsk** [comment]
A little annoying dealing with `package-lock.json`, but I think using it outweighs the costs.

From a developer perspective, you haven't made changes to node modules but you see that there is a diff that includes `package-lock.json`, you should just `git checkout package-lock.json` and go from there ☺️
